### ISSUE TITLE
Flaky test

### ISSUE THREAD

**Issue by TheTincho** [prometheus/alertmanager#427]
Debian users are already starting to report bugs :-)

I can confirm this one. I have seen it once or twice and never again.

This seems like a timer is firing a few µs too early, or a rounding error?

```
The arm64 build of prometheus-alertmanager failed:

  === RUN   TestAggrGroup
  --- FAIL: TestAggrGroup (1.30s)
    dispatch_test.go:132: received batch to early after 299.982ms
  === RUN   TestInhibitRuleHasEqual
  --- PASS: TestInhibitRuleHasEqual (0.00s)
  === RUN   TestInhibitRuleGC
  --- PASS: TestInhibitRuleGC (0.00s)
  === RUN   TestRouteMatch
  --- PASS: TestRouteMatch (0.00s)
  FAIL
  FAIL  github.com/prometheus/alertmanager  1.314s
```

**Comment by TheTincho** [comment]
Another flaky test due to timing issues: `TestBatching` sometimes complains it received 7 instead of 12 alerts. It seems it is highly dependent on timing, so I am disabling it in Debian.

**Comment by TheTincho** [comment]
Found two more flaky tests: TestResolved and TestResolvedFilter, which you can see in this build log: https://buildd.debian.org/status/fetch.php?pkg=prometheus-alertmanager&arch=s390x&ver=0.5.1%2Bds-2&stamp=1481542495

**Comment by TheTincho** [comment]
The timing issues in `dispatch_test.go` get pretty serious under contention. I had already many build failures due to this, and went on to increase the tolerance to 10%, which seems to defeat the purpose of that test. Sadly, I can't find the root cause for these discrepancies.

Also, today I got another build failure test due to the test server not being up fast enough in the acceptance tests. I produced a patch that I think could be merged, I will send a PR.

**Comment by simonpasquier** [comment]
I'm having a look at it since I've experienced the same problems while submitting a PR. I think I can at least explain the problem with `TestBatching` (has `repeat_interval = 5s` and `group_interval = 1s`):
* The dispatcher flushes the alerts at `t0`, sends the notification and marks the notification log at `t0+epsilon`.
* The dispatches flushes the alerts at `t1`, `t2`, `t3` and `t4` and doesn't send the notifications as expected.
* At `t5`, the dispatcher flushes the alerts but then `current_time - (t0+epsilon)` is a bit less then `repeat_interval` so nothing is sent.

Now the notification should be sent again at `t6` for sure.

I've got some logs to make it easier to understand
```
level=debug ts=2018-01-11T12:00:24.45024573Z caller=dispatch.go:430 component=dispatcher aggrGroup={}:{} msg=Flushing alerts="[test4[5afb0ff][active] test1[5ae90ff][active] test5[5af70ff][active] test2[5ae50ff][active] test3[5ae30ff][active]]"
level=debug ts=2018-01-11T12:00:24.461993592Z caller=nflog.go:360 component=nflog msg="Log notification" key={}:{}:default/webhook/0 now=2018-01-11T12:00:24.461890227Z
level=debug ts=2018-01-11T12:00:25.451531089Z caller=dispatch.go:430 component=dispatcher aggrGroup={}:{} msg=Flushing alerts="[test5[5af70ff][active] test2[5ae50ff][active] test3[5ae30ff][active] test4[5afb0ff][active] test1[5ae90ff][active]]"
level=debug ts=2018-01-11T12:00:25.454169134Z caller=notify.go:489 component=dispatcher now=2018-01-11T12:00:25.454145667Z
level=debug ts=2018-01-11T12:00:26.452965629Z caller=dispatch.go:430 component=dispatcher aggrGroup={}:{} msg=Flushing alerts="[test3[5ae30ff][active] test4[5afb0ff][active] test1[5ae90ff][active] test5[5af70ff][active] test2[5ae50ff][active]]"
level=debug ts=2018-01-11T12:00:26.45482201Z caller=notify.go:489 component=dispatcher now=2018-01-11T12:00:26.454793794Z
level=debug ts=2018-01-11T12:00:27.453882226Z caller=dispatch.go:430 component=dispatcher aggrGroup={}:{} msg=Flushing alerts="[test2[5ae50ff][active] test3[5ae30ff][active] test4[5afb0ff][active] test1[5ae90ff][active] test5[5af70ff][active]]"
level=debug ts=2018-01-11T12:00:27.456077197Z caller=notify.go:489 component=dispatcher now=2018-01-11T12:00:27.456054289Z
level=debug ts=2018-01-11T12:00:28.455070086Z caller=dispatch.go:430 component=dispatcher aggrGroup={}:{} msg=Flushing alerts="[test4[5afb0ff][active] test1[5ae90ff][active] test5[5af70ff][active] test2[5ae50ff][active] test3[5ae30ff][active]]"
level=debug ts=2018-01-11T12:00:28.456412435Z caller=notify.go:489 component=dispatcher now=2018-01-11T12:00:28.456398188Z
level=debug ts=2018-01-11T12:00:29.456156537Z caller=dispatch.go:430 component=dispatcher aggrGroup={}:{} msg=Flushing alerts="[test3[5ae30ff][active] test4[5afb0ff][active] test1[5ae90ff][active] test5[5af70ff][active] test2[5ae50ff][active]]"
level=debug ts=2018-01-11T12:00:29.457481007Z caller=notify.go:489 component=dispatcher now=2018-01-11T12:00:29.457462848Z
```

The notification log's timestamp is `2018-01-11T12:00:24.461890227Z` while the last dispatcher's evaluation is at `2018-01-11T12:00:29.457462848Z` and the difference between the two is 4.995572621sec.

**Comment by stuartnelson3** [comment]
closed by #1193
